src/QsCompiler/Tests.Compiler/TestCases/CycleValidation.qs

```diff
@@ -0,0 +1,130 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+
+// Cycle with Generic Resolution
+namespace Microsoft.Quantum.Testing.CycleDetection {
+
+    operation Foo<'A>() : Unit {
+        Bar<'A>();
+    }
+
+    operation Bar<'A>() : Unit {
+        Baz<'A>();
+    }
+
+    operation Baz<'A>() : Unit {
+        Foo<'A>();
+    }
+}
+
+// =================================
+
+// Cycle with Concrete Resolution
+namespace Microsoft.Quantum.Testing.CycleDetection {
+
+    operation Foo<'A,'B,'C>() : Unit {
+        Bar<Int,'B,'C>();
+    }
+
+    operation Bar<'A,'B,'C>() : Unit {
+        Baz<'A,Double,'C>();
+    }
+
+    operation Baz<'A,'B,'C>() : Unit {
+        Foo<'A,'B,String>();
+    }
+}
+
+// =================================
+
+// Constricting Cycle
+namespace Microsoft.Quantum.Testing.CycleDetection {
+
+    operation Foo<'A,'B>() : Unit {
+        Bar<'A,'B>();
+    }
+
+    operation Bar<'A,'B>() : Unit {
+        Baz<'A,'A>();
+    }
+
+    operation Baz<'A,'B>() : Unit {
+        Foo<'A,'B>();
+    }
+}
```

Comment:
Why are constricting cycles not allowed?

Reply:
I believe this was a requirement I got earlier on from Bettina. @bettinaheim Could you comment on why constricting cycles are disallowed?

Reply:
I can modify the TypeParamCombinations tool, which generally encapsulates the rules for how type parameter resolutions are combined, to allow constriction mappings, but then we allow constrictive type arguments within the same method.

Reply:
... or more accurately, _the tool can't be used to disallow_ constrictive type arguments within the same method.

Reply:
I believe one of the reasons for this restriction is that, currently, I don't believe the system has a way of differentiating between a type parameter reference from the previous instance of the callable in the cycle and the type parameters of the current instance of the callable, so there is not currently a way to differentiate between whether a callable is self-constricting or constricting due to the cycle, as far as type parameters are concerned. This is my understanding anyway. Because we don't allow self-constricting callables, we are conservative here and don't allow constriction due to cycles either. If in the future we are able to make this distinction and make sure any other requirements about cycles are still met, we may relax this restriction without making a breaking change, but, I believe, that would require some major changes elsewhere in the system. I'm not sure if there are other reasons constricting cycles are disallowed as a requirement.

Reply:
OK, the finite cycles restriction makes sense, and I understand that this is a conservative approximation of that restriction. Thanks for the explanation.